### PR TITLE
[NeMo-UX] Add PEFT

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?

Initial PR for PEFT in nemo 2.0

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/adapter/base.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        # Load the main module state dict
+        self.to_wrap.load_state_dict(state_dict, strict)
+
+        # Load the adapter module state dict if present
+        if adapter_state_dict:
+            self.adapter.load_state_dict(adapter_state_dict, strict)
+
+
+@dataclass
```
Comment (marcromeyn) [inline nemo/collections/llm/adapter/base.py:54]: No need to make the base-class a dataclass. I think it would be better to include the logic inside the class instead of having to rely on `peft_model_transform`. Something like:

```python
class PEFT:
    def transform(self, module, name=None, prefix=None):
        raise NotImplementedError("The transform method should be implemented by subclasses.")
    
    def __call__(self, model: nn.Module) -> nn.Module:
        model.freeze()
        model.walk(self.transform)
        
        return model
        
class LoRA(PEFT):
    def transform(self, module, name=None, prefix=None):
        ...
```

File: nemo/collections/llm/adapter/lora.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            linear_output, layernorm_output = linear_output
+            adapter_output = self.adapter(layernorm_output)
+        else:
+            adapter_output = self.adapter(x)
+        return linear_output + adapter_output, bias
+
+
+@dataclass
+class LoRAConfig(PEFTConfig):
+    target_modules: List[str] = field(
```
Comment (marcromeyn) [inline nemo/collections/llm/adapter/lora.py:24]: Wouldn't it make more sense to make it batteries-included by add a sensible default for the target_modules?

File: nemo/collections/llm/adapter/lora.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        if isinstance(linear_output, tuple) and len(linear_output) == 2:
+            linear_output, layernorm_output = linear_output
+            adapter_output = self.adapter(layernorm_output)
+        else:
+            adapter_output = self.adapter(x)
+        return linear_output + adapter_output, bias
+
+
+@dataclass
+class LoRAConfig(PEFTConfig):
```
Comment (marcromeyn) [inline nemo/collections/llm/adapter/lora.py:23]: We should add a descriptive doc-string here, including example(s) how to use it

File: nemo/collections/llm/gpt/model/mistral_7b.py [hunk 4]
```diff
@@ -46,8 +46,11 @@ def __init__(
         config: Annotated[Optional[Mistral7BConfig], Config[Mistral7BConfig]] = None,
         optim: Optional[OptimizerModule] = None,
         tokenizer: Optional["TokenizerSpec"] = None,
+        model_transform: Callable = None,
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/model/mistral_7b.py:49]:
```suggestion
        model_transform: Callable[[nn.Module], nn.Module] = None,
```

**Review by marcromeyn — APPROVED** [review]

**Review by marcromeyn — DISMISSED** [review]

**Review by marcromeyn — DISMISSED** [review]

**Review by sararb — DISMISSED** [review]

**Review by sararb — APPROVED** [review]

**Review by sararb — APPROVED** [review]
